File: var_and_let.js
var x = 5; 

if(x == 5){
    var x = 10;
    console.log(x);
}

console.log(x);

// var only supports Global Scope and functional scope it does not support block scope


// But let supports all scopes including block scope
let y = 2;

if(y == 2){
    let y = 3;
    console.log(y);
}

console.log(y);


function changeX(){
    var x = 6;
    console.log(x);
}

function changeY(){
    let y = 6;
    console.log(y);
}

changeX();
console.log('x', x);

changeY();
console.log('y', y);



// Hoisting: Both var and let declarations are hoisted, but there's a difference in how they behave:
// javascript
// Copy code
// console.log(z); // Output: undefined
// var z = 5;
// console.log(z); // Output: 5
// In the above code, the var declaration is hoisted to the top of its scope (global or function), but its assignment isn't. So, console.log(z) doesn't throw an error, but it logs undefined.
// javascript
// Copy code
// console.log(w); // Throws ReferenceError: Cannot access 'w' before initialization
// let w = 5;
// console.log(w);
// In this example, the let declaration is also hoisted, but unlike var, it's not initialized. Accessing w before its declaration throws a ReferenceError.
// Re-declaration: var allows re-declaration within the same scope, whereas let doesn't:
// javascript
// Copy code
// var a = 5;
// var a = 10; // No error
// console.log(a); // Output: 10
// javascript
// Copy code
// let b = 5;
// let b = 10; // Throws SyntaxError: Identifier 'b' has already been declared
// console.log(b);
// Temporal Dead Zone (TDZ): Variables declared with let are subject to the temporal dead zone (TDZ), which means you can't access them before their declaration point:
// javascript
// Copy code
// console.log(c); // Throws ReferenceError: Cannot access 'c' before initialization
// let c = 5;
// While with var, the same code would log undefined.
// Global Object Property: Variables declared with var in the global scope become properties of the global object (window in browsers, global in Node.js), whereas variables declared with let do not:
// javascript
// Copy code
// var d = 5;
// console.log(window.d); // Output: 5

// let e = 10;
// console.log(window.e); // Output: undefined
// These differences highlight the benefits of using let over var in modern JavaScript code, especially when it comes to block scoping, avoiding hoisting-related issues, and preventing accidental re-declarations.

File: var_and_let.cpp
var teacher = "Sanket"; // global
function fun() { // global
    console.log(teacher); // no error will be thrown and it gives undefined
    // console.log(content); // ReferenceError: Cannot access 'content' before  initialization
    // This concept is called temperal dead zone ( TDZ ). The reason before the declaration of a variable which is having a block scope made by let is called as temperal dead  zone.
    // That means you can't access the variable that is having block wcope before it has been declared.
    // 
    
    var teacher = "Pulkit"; // scope of fun
    let content = "JS"; // content will be access only post declaration
    if(content == "JS"){
        let hours = "120+"; 
        console.log(content,hours);
    }
    //  console.log(hours);  // This hours variable is only accessible inside this block scope because the variable declared with the help of let follows the block scope. Error: ReferenceError: hours is not defined
    console.log(teacher, content);
}

fun();
console.log(teacher);

// Difference between function scope and block scope
// If a variable having function scope it is available everywhere in the function.
// For example var is having a function scope so it is available everywhere inside the function even before the declaration of the variable
// But block scope variables are not allowed to do that so if a variable is having a block scope then n that block only it is accessible and it is accessible post declaration only so content will only be accessible inside that block only and it will be available post declaration only.
// For4 example if we want to access hours outside the block it will throw an Error: ReferenceError: hours is not defined 



// Wrap up
// Function scope --> If a variable is available in function scope then is available throughout the function anywhere even before its declaration, so var gives function scope, so it will not give any error in the function even before its declaration, if we used the variable before declaration then it will give undefined

// Block Scope --> if we have anywhere pairs of curly braces it technically defines a block. so if you allow a variable to get a block scope then it will be only available in that block., it won't be available outside the that block and it will be available post declaration only. You can't access it before declarption.
// For example content is only post declaration it..
// So block scope gives the access the variable to a particular block and everything inside that block.and it doesn't give access outside the block.and it will give you access post declaration that is the basic difference between var and let.
// var gives you functional scope and let gives you block scope

